### PR TITLE
Use millisecond time resolution for lingering tasks

### LINKED ISSUE #299: test_max_runtime_exceeded sometimes fails the reason code assertion
Example here:

https://travis-ci.org/twosigma/Cook/jobs/223179765#L1393-L1398

I can't seem to find any usages of reason `:mesos-command-executor-failed` (or 99003) in the scheduler. @wyegelwel, do you have any ideas as to what might cause this?

### PR DESCRIPTION
**Pull request by DaoWen**
Closes #299.

### CODE REVIEW

File: scheduler/src/cook/mesos/scheduler.clj [hunk 1]
```diff
@@ -1059,25 +1058,23 @@
...
-            [?i :instance/task-id ?task-id]
-            [?i :instance/start-time ?start-time]
-            [?j :job/instance ?i]
-            [(get-else $ ?j :job/max-runtime ?default-runtime) ?max-runtime]]
-          db (-> default-timeout-hours time/hours time/in-millis))
-       (keep (fn [[task-id start-time max-runtime]]
-               ;; The convertion between random time units is because time doesn't like
-               ;; Long and Integer/MAX_VALUE is too small for milliseconds
-               (let [timeout-minutes (min (.toMinutes TimeUnit/MILLISECONDS max-runtime)
-                                          (.toMinutes TimeUnit/HOURS max-timeout-hours))]
```
Comment (DaoWen) [inline scheduler/src/cook/mesos/scheduler.clj]: @wyegelwel - This code was rounding any sub-minute values of `:max-runtime` to zero minutes, which was causing the `test_max_runtime_exceeded` integration test to fail nondeterministically. I don't understand the limitation described in the comment above, so I wanted to double-check with you to make sure that by switching to milliseconds I'm not breaking something that's not currently covered by our tests.

**Comment by DaoWen** [comment]
@shamsimam - I updated an existing unit test to also test the sub-minute resolution.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 2]
```diff
@@ -1052,30 +1051,36 @@
                 (reconcile-jobs conn)
                 (reconcile-tasks (db conn) driver framework-id fenzo)))))
 
+;; Unfortunately, clj-time.core/millis only accepts ints, not longs.
+;; The Period class has a constructor that accepts "long milliseconds",
+;; but since that isn't exposed through the clj-time API, we have to call it directly.
+(defn millis->period
```
Comment (DaoWen) [inline scheduler/src/cook/mesos/scheduler.clj]: @wyegelwel - Found the problem! I didn't realize that Clojure was dynamically coercing my longs to ints...

**Comment by wyegelwel** [comment]
That's a... feature? Good find!

**Comment by DaoWen** [comment]
@shamsimam - I feel good about merging this now that I've added the extra unit test scenarios with a max-runtime duration exceeding `Integer/MAX_VALUE` milliseconds.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 1 above)
Comment (DaoWen) [reply scheduler/src/cook/mesos/scheduler.clj]: @shamsimam - Wil and I had a chat and resolved this. See the comment on `millis->period` below for more details.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 3]
```diff
@@ -1052,30 +1051,36 @@
...
+             :where
+             [(ground [:instance.status/unknown :instance.status/running]) [?status ...]]
+             [?i :instance/status ?status]
+             [?i :instance/task-id ?task-id]
+             [?i :instance/start-time ?start-time]
+             [?j :job/instance ?i]
+             [(get-else $ ?j :job/max-runtime ?default-runtime) ?max-runtime]]
+           db (-> default-timeout-hours time/hours time/in-millis))]
+    (for [[task-id start-time max-runtime-ms] jobs-with-max-runtime
+          :let [timeout-ms (millis->period (min max-runtime-ms max-allowed-timeout-ms))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Please rename to `timeout-period`

File: scheduler/src/cook/mesos/scheduler.clj [hunk 4]
```diff
@@ -1052,30 +1051,36 @@
...
-       (keep (fn [[task-id start-time max-runtime]]
-               ;; The convertion between random time units is because time doesn't like
-               ;; Long and Integer/MAX_VALUE is too small for milliseconds
-               (let [timeout-minutes (min (.toMinutes TimeUnit/MILLISECONDS max-runtime)
-                                          (.toMinutes TimeUnit/HOURS max-timeout-hours))]
-                 (when (time/before?
-                         (time/plus (tc/from-date start-time) (time/minutes timeout-minutes))
-                         now)
-                   task-id))))))
+  (let [max-allowed-timeout-ms (-> max-timeout-hours time/hours time/in-millis)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Nitpick: please move this below the query as it isn't used in the query.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 2 above)
Comment (shamsimam) [reply scheduler/src/cook/mesos/scheduler.clj]: Nitpick: we can make this function private.

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 5]
```diff
@@ -423,38 +423,55 @@
...
+                                   :user "tsram"
+                                   :job-state :job.state/running
+                                   :max-runtime 10000)
+        instance-id-4 (create-dummy-instance conn job-id-4
+                                             :start-time #inst "2015-01-05")
+
+        job-id-5 (create-dummy-job conn
+                                   :user "tsram"
+                                   :job-state :job.state/running
+                                   ; timeout value exceeds Integer/MAX_VALUE millis
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: +1

**Comment by DaoWen** [comment]
@shamsimam - Feedback addressed.

**Pull request merged**